### ISSUE TITLE
Establish a set of good  measures and missing metrics

### ISSUE THREAD

**Issue by BenJam** [librariesio/libraries.io#1486]
## Help Wanted
This ticket is an open for comment and collaboration.

## Goals
* To help others make more informed decisions about software they include in their own projects. 
* To define a set of metrics that are missing from Libraries.io
* To define a set of useful direct and derivative measures for judging aspects of a project (i.e. code quality, community, documentation).
* To refine Libraries.io's own 'SourceRank' measure ([example](https://libraries.io/npm/rails/sourcerank))

## Examples of similar work
* Linux Foundation's census project https://github.com/linuxfoundation/cii-best-practices-badge
* IILab's Open Integrity Index https://github.com/openintegrity (focusses on end user applications rather than libraries)

## [Questions](https://gist.github.com/BenJam/9e1d6276efabced6e3973137e6993b01)
For me this process begins with a number of framing questions. These questions are user-centred and based on the needs of our users, as defined in our [personas](https://docs.libraries.io/personas.html) from there we can define measures and metrics. 

## [Measures](https://gist.github.com/BenJam/fa356768c0d6a4949493a6ccf2d7eec1)
Measures can be direct (a quoted metric i.e. 'released on 1st Jan 2017), derivative (information gleaned from data i.e. 'released more than a year ago') or aggregated (compiled from >1 data i.e. 'all releases we're within the last year'). 

Measures may be broken down into a number of areas. At Libraries.io we have been considering areas for code, community, distribution and documentation. The single, overarching measure in Libraries.io is SourceRank which is defined over [in our documentation](https://docs.libraries.io/overview#sourcerank). 

## [Metrics](https://gist.github.com/BenJam/7c9e38d07d36cc78cfa70a6e3fbda53e)
Metrics can be 'harvested' (gathered automatically using APIs, data dumps, scraping etc) or 'farmed' (gathered from a community by contribution).

Libraries.io currently focusses on harvesting metrics. It is preferable for a metric to be present in many sources rather than a single source so that we can make like-for-like comparisons across [supported package managers](http://libraries.io/compatibility). 

## Sources
Sources contain metrics. They may also contain measures themselves.  We think it is important not to rely on proprietary, third party services for measures. 

Reproducibility is the key issue here. An inability to reproduce a measure from the source data (metrics) risks the ability to create a like for like comparison of two pieces of software and ties all users of the classifier to that service. This is [unacceptable](https://www.youtube.com/watch?v=07So_lJQyqw)(🔊).
 
## Contributing 
While there will be no one single approach that is right for everyone our hope is that we can come to consensus about what good measures look like and what metrics they will require, so that Libraries.io can provide them. 

Please feel free to write up your own thoughts, ideas or draft specifications for areas, measures etc. and reference that work here in a comment but be sure to state very clearly:

* what metric(s) are missing from Libraries.io
* which sources and/or collection methods you envisage using to gather this data

**Comment by BenJam** [comment]
For ref: IILAB's thinking on project metrics: https://github.com/openintegrity/openintegrity-metrics/blob/master/METRICS.md

**Comment by BenJam** [comment]
Ossert has established a number of classifiers under the areas of Maintenance, Popularity and Maturity. https://ossert.evilmartians.io/lalala-development

**Comment by sclinede** [comment]
From the Ossert point of view, I want to get rid of the values like A, B and so. Because I think that the best thing that could be done for a user and maintainer is to understand the "group" this project belongs.
For example, it could be "the enterprise level", "experimental", "newborn", "need assistance", "lacks support", "trending", "under development" and so on.

On the other hand, each metric (as I see) should have a level regarding how many projects achieved it.
As I understand it will help us with indexing and also gives some motivation or guide what to seek for.
For example, "Project X belongs to 5% most responsive Ruby projects" or "Project X belongs to 15% most popular Ruby projects".

**Comment by BenJam** [comment]
> On the other hand, each metric (as I see) should have a "level" regarding how many projects achieved it.

This is a fair point. If it's possible to say that this project is within the top x% of projects by this measure then that's useful information, but it won't be applicable to all types of measures. However it is the approach that we've taken with our primary measure no. of dependent projects/repositories, which we've compiled into https://libraries.io/digital-infrastructure 

> As I understand it will help us with indexing and also gives some motivation or guide what to seek for.

This is a very important point and one which we've explored in the past with the Linux Foundation and IILAB, both of whom have badge programmes as their means of motivating maintainers to improve projects. 

I think this could be successful without a badge though. Transparency about *what* is included in the measure and *how* the measure is calculated may be enough to educate users if posed in an intelligible and straightforward way, perhaps with a little direct instruction i.e. "improve your SourceRank by... '
